doc/source/train/examples/intel_gaudi/llama_deepspeed.ipynb

```diff
@@ -0,0 +1,2436 @@
+    "\n",
+    "    # prepare data collator\n",
+    "    data_collator = DataCollatorForLanguageModeling(tokenizer, pad_to_multiple_of=8, return_tensors=\"pt\", mlm=False)\n",
+    "\n",
+    "    gaudi_config = GaudiConfig()\n",
+    "    gaudi_config.use_fused_adam = True\n",
+    "    gaudi_config.use_fused_clip_norm = True\n",
+    "\n",
+    "    trainer = GaudiTrainer(\n",
+    "        model=model,\n",
+    "        gaudi_config=gaudi_config,\n",
+    "        args=training_args,\n",
+    "        train_dataset=tokenized_datasets[\"train\"],\n",
+    "        eval_dataset=None,\n",
+    "        tokenizer=tokenizer,\n",
+    "        data_collator=data_collator,\n",
+    "        compute_metrics=None,\n",
+    "        preprocess_logits_for_metrics=None,\n",
+    "    )\n",
+    "\n",
+    "    train_result = trainer.train()\n",
+    "    print(f\"train_result = {train_result}\")\n",
+    "    trainer.save_model()\n",
+    "\n",
+    "    return train_result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Main Training Function\n",
+    "The `train_llama` function sets up the distributed training environment using Ray and starts the training process. To enable training using HPU, we only need to make the following changes:\n",
+    "- Set the exectuion mode for training, supported execution mode are:\n",
+    "\n",
+    "    - \"lazy\": Deferred execution of graphs, comprising of ops delivered from script op by op similar to Eager mode. It gives the Eager mode experience with performance on Gaudi. Unlike Eager Mode with torch.compile, graph is analyzed in each iteration leading to a higher CPU usage.\n",
+    "    - \"eager\": Op-by-op execution as defined in standard PyTorch Eager mode scripts.\n",
+    "    - \"eager.compile\": Eager mode extended with `torch.compile` - Similar to Eager mode but extended with wrapping complete or part of model (such as a function) into a graph. Parts that are not wrapped are executed eagerly.\n",
+    "\n",
+    "    More detail theory can be found [here](https://docs.habana.ai/en/latest/PyTorch/Reference/PyTorch_Gaudi_Theory_of_Operations.html)\n",
+    "- Require an HPU for each worker in ScalingConfig\n",
+    "- Set backend to `hccl` in TorchConfig"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "def train_llama(num_workers, execution_mode):\n",
+    "    import ray\n",
+    "    from ray.train import ScalingConfig\n",
+    "    from ray.train.torch import TorchTrainer, TorchConfig\n",
+    "\n",
+    "    # deepspeed config, can also place it to config file\n",
+    "    deepspeed_config = {\n",
+    "        \"steps_per_print\": 64,\n",
+    "        \"train_batch_size\": \"auto\",\n",
+    "        \"train_micro_batch_size_per_gpu\": \"auto\",\n",
+    "        \"gradient_accumulation_steps\": \"auto\",\n",
```

Comment:
For curiosity, does `GaudiTrainer` or Deepspeed figure out the actual numbers at runtime?

Reply:
Yes, those option will be calculated when TrainingArguments object instanced.